pandas/core/internals/blocks.py

```diff
@@ -798,14 +801,27 @@ def replace(
             filtered_out = ~self.mgr_locs.isin(filter)
             mask[filtered_out.nonzero()[0]] = False
 
+        if not mask.any():
+            if inplace:
+                return [self]
+            return [self.copy()]
+
         try:
             blocks = self.putmask(mask, value, inplace=inplace)
+            # Note: it is _not_ the case that self._can_hold_element(value)
+            #  is always true at this point.  In particular, that can fail
+            #  for:
+            #   "2u" with bool-dtype, float-dtype
+            #   0.5 with int64-dtype
+            #   np.nan with int64-dtype
         except (TypeError, ValueError):
             # GH 22083, TypeError or ValueError occurred within error handling
             # causes infinite loop. Cast and retry only if not objectblock.
             if is_object_dtype(self):
                 raise
 
+            assert not self._can_hold_element(value), value
+
```

Comment:
Tests?  Does this bubble up?

Reply:
This is hit in tests.  The original hope was to replace the try/except a few lines above with a `if self._can_hold_element(value):` but as noted in comments a few lines up, the try works in some cases where _can_hold_element fails